run_target/DEPENDS_KW: accept CustomTargetIndex

gerion0 opened the pull request:
Code like
```meson
a = custom_target(..., output: 'foo', 'bar')
foo = a[0]

run_target(..., depends: foo)
```
does not setup otherwise.

gerion0 commented:
BTW, do you plan to generate the documentation out of the Python types directly (it would combine all typing information in one place)? The whole approach seems related to the tool that we used for checking the types in #9646. We build a typing hierarchy for Meson directly in Python there, too ([for example](https://github.com/nnamua/meson-classification-checker/blob/main/functions.py#L83)).

gerion0 commented:
Sadly, this bugfix does not made it in 1.4.0. Would be nice, to have it in 1.4.1, if you release such a version.

bruchar1 commented on mesonbuild/interpreter/type_checking.py:
```diff
@@ -269,9 +269,9 @@ def _env_convertor(value: _FullEnvInitValueType) -> EnvironmentVariables:
 )
 
 # TODO: CustomTargetIndex should be supported here as well
```
This makes this TODO no longer relevant...

bruchar1 reviewed (commented):
I think you need to change the types for for `dependencies` kwarg of `RunTarget` as well, and for the `get_dependencies()` method of this class.

dcbaker commented on mesonbuild/interpreter/type_checking.py:
```diff
@@ -269,9 +269,9 @@ def _env_convertor(value: _FullEnvInitValueType) -> EnvironmentVariables:
-    ContainerTypeInfo(list, (BuildTarget, CustomTarget)),
+    ContainerTypeInfo(list, (BuildTarget, CustomTarget, CustomTargetIndex)),
     listify=True,
     default=[],
```
```suggestion
    default=[],
    since_types={CustomTargetIndex: '1.5.0'},
```

gerion0 replied:
I guess, you meant `since_values`. This is added now.

dcbaker commented on docs/yaml/functions/run_target.yaml:
```diff
@@ -40,7 +40,7 @@ kwargs:
 
   depends:
-    type: list[build_tgt | custom_tgt]
+    type: list[build_tgt | custom_tgt | custom_idx]
```
This needs a note at the end of this block that `custom_idx` support was added in 1.5.0

gerion0 replied:
Actually, it was possible before the typing checks (0.60). I have documented it this way.

dcbaker reviewed (changes requested):
I'd really love to see a test to make sure this works, If there's already a test that takes a `custom_target` as a dependency just adding a second case to that test that converts (for example) `depends : dependtarget` -> `depends : dependentarget[0]` would be sufficient.

gerion0 commented:
Thanks for the review. I should have addressed all comments. I additionally fixed the documentation in all cases where `DEPENDS_KW` is used.

Some random notes:
- The Gnome module documentation already documents `CustomTargetIndex` while it currently shouldn't be possible.
- The `test()` meson function has a `depends` KW while not using `DEPENDS_KW` in the source code.
- I did not found a test case that tests a `generator` that depends on a `custom_target` (the closest was a dependency to an executable in `test\ cases/native/3\ pipeline/depends/meson.build`).

gerion0 commented:
Hmm, sorry, the diff between the two commits is broken since I also rebased to the current master. Should I add following changes as extra commits and squash it in the end or keep to force push?

bruchar1 commented:
You still need to adjust type annotations in build.py, for `RunTarget`...

gerion0 commented:
> You still need to adjust type annotations in build.py, for `RunTarget`...

I'm not sure what you mean precisely. Do you think of something like that:
```diff
diff --git a/mesonbuild/build.py b/mesonbuild/build.py
index 2b6ce8875..93c5125ab 100644
--- a/mesonbuild/build.py
+++ b/mesonbuild/build.py
@@ -2834,7 +2834,7 @@ class RunTarget(Target, CommandBase):
         repr_str = "<{0} {1}: {2}>"
         return repr_str.format(self.__class__.__name__, self.get_id(), self.command[0])
 
-    def get_dependencies(self) -> T.List[T.Union[BuildTarget, 'CustomTarget']]:
+    def get_dependencies(self) -> T.List[T.Union[BuildTarget, 'CustomTarget', 'CustomTargetIndex']]:
         return self.dependencies
 
     def get_generated_sources(self) -> T.List['GeneratedTypes']:
```

bruchar1 reviewed (approved):
LGTM!

dcbaker commented:
Failures look relavent

gerion0 commented:
> Failures look relavent

They are. Since I do not own a Windows system, I need to leverage the pipeline here. The change hopefully fixes the error.

gerion0 commented:
I'm not getting the new errors. They are different but Windows specific. The best error message I found was this (with no reason statet for the failure):
```
[6/34] "cl" "-Ilibc_wrapper.a.p" "-I." "-I..\test cases\rust\12 bindgen" "/MDd" "/nologo" "/showIncludes" "/utf-8" "/W2" "/Od" "/Zi" "-DPROJECT_ARG" "/fp:fast" "-DGLOBAL_ARG" "/fp:fast" "-DCMD_ARG" "/Fdlibc_wrapper.a.p\meson-generated_.._header3.c.pdb" /Folibc_wrapper.a.p/meson-generated_.._header3.c.obj "/c" header3.c
FAILED: libc_wrapper.a.p/meson-generated_.._header3.c.obj
```
However, the Python stack traces from above that clearly resulted of my changes are fixed.

bruchar1 commented on mesonbuild/backend/vs2010backend.py:
```diff
@@ -327,6 +327,9 @@ def get_obj_target_deps(self, obj_list):
     def get_target_deps(self, t: T.Dict[T.Any, build.Target], recursive=False):
```
```suggestion
    def get_target_deps(self, t: T.Dict[T.Any, T.Union[build.Target, build.CustomTargetIndex]], recursive=False):
```

gerion0 replied:
Sure, missed that. Fixed now.

bruchar1 commented:
> I'm not getting the new errors. They are different but Windows specific. 

The rust bindgen test failure seems irrelevant. I saw the same failure on another unrelated PR.

dcbaker commented:
Bindgen on windows seems to fail sporadically, seems like the version in the base image sometimes doesn’t work
